### PR TITLE
Implements docker firewall validation

### PR DESCRIPTION
**Pull request by b10n1k**
I added a small validation for the expected configuration of the firewall.
Briefly, makes sure that Docker firewall zone exists with an active docker0
interface and in turn this is not added on trusted zone. Run a container and
verify the iptables.

Applies firewall test in docker and docker_image
the tests implemented based on https://docs.docker.com/network/iptables/

- Related ticket: https://progress.opensuse.org/issues/89872
- Verification run: 
opensuse-Tumbleweed-DVD-x86_64-Build20210508-extra_tests_textmode_containers@64bit -> https://openqa.opensuse.org/tests/1735159
opensuse-Tumbleweed-DVD-x86_64-Build20210508-extra_tests_textmode_containers@64bit -> https://openqa.opensuse.org/tests/1735160
opensuse-15.3-DVD-x86_64-Build148.1-extra_tests_textmode_containers@64bit -> https://openqa.opensuse.org/tests/1735161
sle-15-SP3-Online-x86_64-Build187.1-containers_basic@64bit -> https://openqa.suse.de/tests/5996960
sle-15-SP3-Online-x86_64-Build187.1-containers_sle_image_on_sle_host@64bit -> https://openqa.suse.de/tests/5996961
sle-15-SP3-Online-x86_64-Build187.1-containers_sle_image_on_centos_host@64bit -> https://openqa.suse.de/tests/5996962
sle-15-SP3-Online-x86_64-Build187.1-containers_sle_image_on_opensuse_host@64bit -> https://openqa.suse.de/tests/5996963
sle-15-SP3-Online-x86_64-Build187.1-containers_sle_image_on_ubuntu_host@64bit -> https://openqa.suse.de/tests/5996964
sle-15-SP2-Server-DVD-Updates-x86_64-Build20210511-1-mau-extratests-docker@64bit -> https://openqa.suse.de/tests/5997086
sle-15-SP1-Server-DVD-Updates-x86_64-Build20210511-1-mau-extratests-docker@64bit -> https://openqa.suse.de/tests/5996966
sle-15-Server-DVD-Updates-x86_64-Build20210511-1-mau-extratests-docker@64bit -> https://openqa.suse.de/tests/5996967
sle-12-SP5-Server-DVD-Updates-x86_64-Build20210511-1-mau-extratests-docker@64bit -> https://openqa.suse.de/tests/5996959
sle-12-SP4-Server-DVD-Updates-x86_64-Build20210511-1-mau-extratests-docker@64bit -> https://openqa.suse.de/tests/5996968
sle-12-SP3-Server-DVD-Updates-x86_64-Build20210510-2-mau-extratests-docker@64bit -> https://openqa.suse.de/tests/5996969

### CODE REVIEW

File: lib/containers/utils.pm [hunk 1]
```diff
@@ -41,20 +43,65 @@ sub test_seccomp {
...
+sub check_docker_firewall {
+    my $container_name = 'sut_container';
+    my $docker_version = get_docker_version();
+    systemctl('is-active firewalld');
+    my $running = script_output qq(docker ps -q | wc -l);
+    validate_script_output('ip a s docker0', sub { /state DOWN/ }) if $running == 0;
+    assert_script_run "firewall-cmd --list-all --zone=docker";
+    # Docker zone is created only for SLE15-SP3 for version <20.10
+    validate_script_output "firewall-cmd --list-interfaces --zone=docker", sub { /docker0/ } if (check_runtime_version($docker_version, ">=20.10") || is_sle("15-SP3+"));
+    validate_script_output "firewall-cmd --list-interfaces --zone=trusted", sub { /^\s*$/ };
```
Comment (pdostal) [inline lib/containers/utils.pm:55]: What this line does?

**Comment by pdostal** [comment]
From my side this looks nice. Just please address @asmorodskyi's suggestions from #12480

File: lib/containers/utils.pm (hunk 1 above)
Comment (b10n1k) [reply lib/containers/utils.pm:55]: Same thing asks @asmorodskyi . https://docs.docker.com/network/iptables/#integration-with-firewalld. just another check

**Comment by b10n1k** [comment]
> From my side this looks nice. Just please address @asmorodskyi's suggestions from #12480

i have the impression i done all except the `version_utils::check_version`

File: lib/containers/utils.pm (hunk 1 above)
Comment (pdostal) [reply lib/containers/utils.pm:55]: I found this [question](https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/12480#discussion_r627278432) from Anton:
> what this line suppose to prove ? that some interfaces are in trusted ? how it is related to docker + firewalld ?

I read your link but I still have the same question :)

File: lib/containers/utils.pm [hunk 2]
```diff
@@ -41,20 +43,65 @@ sub test_seccomp {
...
+    my $container_ip = container_ip($container_name, "docker");
+    # Each running container should have added a new entry to the DOCKER zone.
+    # ex. output: "1           0        0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.2           tcp dpt:1234"
+    validate_script_output "iptables -L DOCKER -nvx --line-numbers", sub { /1.+ACCEPT.+!docker0 docker0.+$container_ip\s+tcp dpt:1234/ };
+    assert_script_run "docker kill $container_name ";
+}
+
+sub get_docker_version {
+    my $v = script_output("docker --version");
+    record_info "$v", $v =~ /(\d{2}\.\d{2})/;
```
Comment (pdostal) [inline lib/containers/utils.pm:69]: As Anton suggested below let's just increase the verbosity for reviewers:
```suggestion
    $v =~ /(\d{2}\.\d{2})/;
    record_info "$v", "Docker version $v"; 
```

**Review by asmorodskyi — APPROVED** [review]
LGTM

File: lib/containers/utils.pm (hunk 1 above)
Comment (b10n1k) [reply lib/containers/utils.pm:55]: maybe i am wrong but Seems that when firewalld was used in some cases you need to have the docker0 in trusted zone to have internet access. based on the documentation and the default setup i try to ensure that docker0 is not set for that zone. i could also check all the zones but i thought it was not necessary for the others

Comment (b10n1k) [reply lib/containers/utils.pm:55]: by default? not if not someone added IMO.

File: lib/containers/utils.pm [hunk 3]
```diff
@@ -41,20 +43,65 @@ sub test_seccomp {
...
+    validate_script_output "firewall-cmd --list-interfaces --zone=docker", sub { /docker0/ } if (check_runtime_version($docker_version, ">=20.10") || is_sle('>=15-sp3'));
+    validate_script_output "firewall-cmd --list-interfaces --zone=trusted", sub { /^\s*$/ };
+    # Rules applied before DOCKER. Default is to listen to all tcp connections
+    # ex. output: "1           0        0 RETURN     all  --  *      *       0.0.0.0/0            0.0.0.0/0"
+    validate_script_output "iptables -L DOCKER-USER -nvx --line-numbers", sub { /1.+all.+0\.0\.0\.0\/0\s+0\.0\.0\.0\/0/ };
+    assert_script_run "docker run -id --rm --name $container_name -p 1234:1234 " . registry_url('alpine');
+    my $container_ip = container_ip($container_name, "docker");
+    # Each running container should have added a new entry to the DOCKER zone.
+    # ex. output: "1           0        0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.0.2           tcp dpt:1234"
+    validate_script_output "iptables -L DOCKER -nvx --line-numbers", sub { /1.+ACCEPT.+!docker0 docker0.+$container_ip\s+tcp dpt:1234/ };
```
[the remark this answers is not in the document]
Comment (b10n1k) [reply lib/containers/utils.pm:62]: thanks a lot. this was a good one to know. but i agree with @asmorodskyi to reserve an implementation for the future

**Review by jlausuch — APPROVED** [review]

**Comment by grisu48** [comment]
This one causes [poo#92728](https://progress.opensuse.org/issues/92728) and breaks on publiccloud test runs.
A fix should be easy and I'm on it.

**Comment by Vogtinator** [comment]
Also fails on 15.2: https://openqa.opensuse.org/tests/1744829#step/docker/96

@grisu48 Is that the same failure?

**Comment by asmorodskyi** [comment]
> Also fails on 15.2: https://openqa.opensuse.org/tests/1744829#step/docker/96
> 
> @grisu48 Is that the same failure?

failure in Leap 15.2 is due to usage of docker 19.03 ( no special zone created ) 
failure in public is because test is expecting firewalld to run and it is not running

**Comment by grisu48** [comment]
@Vogtinator as @asmorodskyi said, but we can track both in https://progress.opensuse.org/issues/92728. I'm working on a fix.

**Comment by pdostal** [comment]
The fix Felix works on is #12538.

**Pull request merged**